src/topology/algebra/module/character_space.lean

```diff
@@ -0,0 +1,124 @@
+/-
+Copyright (c) 2022 Frédéric Dupuis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Frédéric Dupuis
+-/
+
+import topology.algebra.module.weak_dual
+import algebra.algebra.spectrum
+
+/-!
+# Character space of a topological algebra
+
+The character space of a topological algebra is the subset of elements of the weak dual that
+are also algebra homomorphisms. This space is used in the Gelfand transform, which gives an
+isomorphism between a commutative C⋆-algebra and continuous functions on the character space
+of the algebra. This, in turn, is used to construct the continuous functional calculus on
+C⋆-algebras.
+
+
+## Implementation notes
+
+We define `character_space 𝕜 A` as a subset of the weak dual, which automatically puts the
+correct topology on the space. We then define `to_alg_hom` which provides the algebra homomorphism
+corresponding to any element. We also provide `to_clm` which provides the element as a
+continuous linear map. (Even though `weak_dual 𝕜 A` is a type copy of `A →L[𝕜] 𝕜`, this is
+often more convenient.)
+
+## TODO
+
+* Prove that the character space is a compact subset of the weak dual. This requires the
+  Banach-Alaoglu theorem.
+
+## Tags
+
+character space, Gelfand transform, functional calculus
+
+-/
+
+namespace weak_dual
+
+/-- The character space of a topological algebra is the subset of elements of the weak dual that
+are also algebra homomorphisms. -/
+def character_space (𝕜 : Type*) (A : Type*) [comm_semiring 𝕜] [topological_space 𝕜]
+  [has_continuous_add 𝕜] [has_continuous_const_smul 𝕜 𝕜]
+  [non_unital_non_assoc_semiring A] [topological_space A] [module 𝕜 A] :=
+  {φ : weak_dual 𝕜 A | (φ ≠ 0) ∧ (∀ (x y : A), φ (x * y) = (φ x) * (φ y))}
+
+variables {𝕜 : Type*} {A : Type*}
+
+namespace character_space
+
+section non_unital_non_assoc_semiring
+
+variables [comm_semiring 𝕜] [topological_space 𝕜] [has_continuous_add 𝕜]
+  [has_continuous_const_smul 𝕜 𝕜] [non_unital_non_assoc_semiring A] [topological_space A]
```

Comment:
`has_continuous_const_smul 𝕜 𝕜` feels like quite a strange typeclass to use. Are there real examples where `has_continuous_const_smul 𝕜 𝕜` is true but `has_continuous_const_mul 𝕜` (which implies it) is not?

Reply:
No idea, but that's what's required in the definition of `weak_dual` so I just went with the same thing.

Reply:
I was just taking the weakest (no pun intended) typeclass that gave me the lemmas I needed, feel free to change it if you like

Reply:
I'm happy to change it if @eric-wieser thinks it's a good idea, but I'm inclined to just leave it like this if it's anyway a weaker assumption.

Reply:
If this is just to match `weak_dual`, then changing this is out of scope for this PR anyway; if we do decide to change it, `weak_dual` should change too.